File: WinForms/.NET Framework/FormDesigner/VBNET/AppForm.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports O2S.Components.PDFView4NET
Imports O2S.Components.PDFView4NET.Forms

Namespace O2S.Samples.PDFView4NET.FormDesigner
	Public Partial Class AppForm
		Inherits Form
		Public Sub New()
			InitializeComponent()
		End Sub

		Private clickedFieldWidget As PDFFieldWidget

        Private Sub AppForm_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
            tscbxZoom.SelectedIndex = 2
        End Sub

		#Region "Toolbar Events"

        Private Sub tsbFileOpen_Click(ByVal sender As Object, ByVal e As EventArgs) Handles tsbFileOpen.Click
            If ofd.ShowDialog() = DialogResult.OK Then
                document.Load(ofd.FileName)
            End If
        End Sub

        Private Sub tsbFileSave_Click(ByVal sender As Object, ByVal e As EventArgs) Handles tsbFileSave.Click
            If sfd.ShowDialog() = DialogResult.OK Then
                document.Save(sfd.FileName)
            End If
        End Sub

        Private Sub tsbPanAndScan_Click(ByVal sender As Object, ByVal e As EventArgs) Handles tsbPanAndScan.Click
            pageView.WorkMode = UserInteractiveWorkMode.PanAndScan
        End Sub

        Private Sub tsbEditFormFields_Click(ByVal sender As Object, ByVal e As EventArgs) Handles tsbEditFormFields.Click
            pageView.WorkMode = UserInteractiveWorkMode.EditAnnotations
        End Sub

        Private Sub tsbFieldTextBox_Click(ByVal sender As Object, ByVal e As EventArgs) Handles tsbFieldTextBox.Click
            pageView.WorkMode = UserInteractiveWorkMode.AddTextBoxField
        End Sub

        Private Sub tsbFieldCheckBox_Click(ByVal sender As Object, ByVal e As EventArgs) Handles tsbFieldCheckBox.Click
            pageView.WorkMode = UserInteractiveWorkMode.AddCheckBoxField
        End Sub

        Private Sub tsbFieldRadioButton_Click(ByVal sender As Object, ByVal e As EventArgs) Handles tsbFieldRadioButton.Click
            pageView.WorkMode = UserInteractiveWorkMode.AddRadioButtonField
        End Sub

        Private Sub tsbFieldListBox_Click(ByVal sender As Object, ByVal e As EventArgs) Handles tsbFieldListBox.Click
            pageView.WorkMode = UserInteractiveWorkMode.AddListBoxField
        End Sub

        Private Sub tsbFieldComboBox_Click(ByVal sender As Object, ByVal e As EventArgs) Handles tsbFieldComboBox.Click
            pageView.WorkMode = UserInteractiveWorkMode.AddDropDownListField
        End Sub

        Private Sub tsbFieldPushButton_Click(ByVal sender As Object, ByVal e As EventArgs) Handles tsbFieldPushButton.Click
            pageView.WorkMode = UserInteractiveWorkMode.AddPushButtonField
        End Sub

        Private Sub tsbFieldDigitalSignature_Click(ByVal sender As Object, ByVal e As EventArgs) Handles tsbFieldDigitalSignature.Click
            pageView.WorkMode = UserInteractiveWorkMode.AddSignatureField
        End Sub

        Private Sub tscbxZoom_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles tscbxZoom.SelectedIndexChanged
            pageView.Zoom = Single.Parse(tscbxZoom.SelectedItem.ToString())
        End Sub

		#End Region

		#Region "PDF Viewer Events"

        Private Sub pageView_WorkModeChanged(ByVal sender As Object, ByVal e As EventArgs) Handles pageView.WorkModeChanged
            tsbPanAndScan.Checked = pageView.WorkMode = UserInteractiveWorkMode.PanAndScan
            tsbEditFormFields.Checked = pageView.WorkMode = UserInteractiveWorkMode.EditAnnotations
            tsbFieldTextBox.Checked = pageView.WorkMode = UserInteractiveWorkMode.AddTextBoxField
            tsbFieldCheckBox.Checked = pageView.WorkMode = UserInteractiveWorkMode.AddCheckBoxField
            tsbFieldRadioButton.Checked = pageView.WorkMode = UserInteractiveWorkMode.AddRadioButtonField
            tsbFieldListBox.Checked = pageView.WorkMode = UserInteractiveWorkMode.AddListBoxField
            tsbFieldComboBox.Checked = pageView.WorkMode = UserInteractiveWorkMode.AddDropDownListField
            tsbFieldPushButton.Checked = pageView.WorkMode = UserInteractiveWorkMode.AddPushButtonField
            tsbFieldDigitalSignature.Checked = pageView.WorkMode = UserInteractiveWorkMode.AddSignatureField
        End Sub

        Private Sub pageView_BeforeFieldAdd(ByVal sender As Object, ByVal e As BeforeFieldAddEventArgs) Handles pageView.BeforeFieldAdd
            Dim cancel As Boolean = False
            Select Case e.Field.Type
                Case PDFFieldType.TextBox
                    Dim tbpf As New TextBoxPropertiesForm()
                    tbpf.TextBoxWidget = TryCast(e.Field.Widgets(0), PDFTextBoxWidget)
                    cancel = tbpf.ShowDialog() = DialogResult.Cancel
                    Exit Select
                Case PDFFieldType.CheckBox
                    Dim cbpf As New CheckBoxPropertiesForm()
                    cbpf.CheckBoxWidget = TryCast(e.Field.Widgets(0), PDFCheckBoxWidget)
                    cancel = cbpf.ShowDialog() = DialogResult.Cancel
                    Exit Select
                Case PDFFieldType.RadioButton
                    Dim rbpf As New RadioButtonPropertiesForm()
                    rbpf.RadioButtonWidget = TryCast(e.Field.Widgets(0), PDFRadioButtonItem)
                    cancel = rbpf.ShowDialog() = DialogResult.Cancel
                    Exit Select
                Case PDFFieldType.PushButton
                    Dim pbpf As New PushButtonPropertiesForm()
                    pbpf.PushButtonWidget = TryCast(e.Field.Widgets(0), PDFPushButtonWidget)
                    cancel = pbpf.ShowDialog() = DialogResult.Cancel
                    Exit Select
                Case PDFFieldType.ListBox
                    Dim lbpf As New ListBoxPropertiesForm()
                    lbpf.ListBoxWidget = TryCast(e.Field.Widgets(0), PDFListBoxWidget)
                    cancel = lbpf.ShowDialog() = DialogResult.Cancel
                    Exit Select
                Case PDFFieldType.DropDownList
                    Dim ddlpf As New DropdownListPropertiesForm()
                    ddlpf.DropDownListWidget = TryCast(e.Field.Widgets(0), PDFDropDownListWidget)
                    cancel = ddlpf.ShowDialog() = DialogResult.Cancel
                    Exit Select
                Case PDFFieldType.Signature
                    Dim spf As New SignaturePropertiesForm()
                    spf.SignatureWidget = TryCast(e.Field.Widgets(0), PDFSignatureWidget)
                    cancel = spf.ShowDialog() = DialogResult.Cancel
                    Exit Select
            End Select

            e.Cancel = cancel
        End Sub

        Private Sub pageView_AfterFieldAdd(ByVal sender As Object, ByVal e As AfterFieldAddEventArgs) Handles pageView.AfterFieldAdd

        End Sub

        Private Sub pageView_FieldContextMenu(ByVal sender As Object, ByVal e As FieldContextMenuEventArgs) Handles pageView.FieldContextMenu
            If pageView.WorkMode = UserInteractiveWorkMode.EditAnnotations Then
                clickedFieldWidget = e.FieldWidget
                cmsFieldContextMenu.Show(pageView.PointToScreen(e.MousePosition))
            End If
        End Sub

		#End Region

		#Region "Context Menu Events"

        Private Sub tsmiFieldContextMenuProperties_Click(ByVal sender As Object, ByVal e As EventArgs) Handles tsmiFieldContextMenuProperties.Click
            Select Case clickedFieldWidget.Field.Type
                Case PDFFieldType.TextBox
                    pageView.BeginContentUpdate()

                    Dim tbpf As New TextBoxPropertiesForm()
                    tbpf.TextBoxWidget = TryCast(clickedFieldWidget, PDFTextBoxWidget)
                    tbpf.ShowDialog()

                    pageView.EndContentUpdate()
                    Exit Select
                Case PDFFieldType.CheckBox
                    pageView.BeginContentUpdate()

                    Dim cbpf As New CheckBoxPropertiesForm()
                    cbpf.CheckBoxWidget = TryCast(clickedFieldWidget, PDFCheckBoxWidget)
                    cbpf.ShowDialog()

                    pageView.EndContentUpdate()
                    Exit Select
                Case PDFFieldType.RadioButton
                    pageView.BeginContentUpdate()

                    Dim rbpf As New RadioButtonPropertiesForm()
                    rbpf.RadioButtonWidget = TryCast(clickedFieldWidget, PDFRadioButtonItem)
                    rbpf.ShowDialog()

                    pageView.EndContentUpdate()
                    Exit Select
                Case PDFFieldType.PushButton
                    pageView.BeginContentUpdate()

                    Dim pbpf As New PushButtonPropertiesForm()
                    pbpf.PushButtonWidget = TryCast(clickedFieldWidget, PDFPushButtonWidget)
                    pbpf.ShowDialog()

                    pageView.EndContentUpdate()
                    Exit Select
                Case PDFFieldType.ListBox
                    pageView.BeginContentUpdate()

                    Dim lbpf As New ListBoxPropertiesForm()
                    lbpf.ListBoxWidget = TryCast(clickedFieldWidget, PDFListBoxWidget)
                    lbpf.ShowDialog()

                    pageView.EndContentUpdate()
                    Exit Select
                Case PDFFieldType.DropDownList
                    pageView.BeginContentUpdate()

                    Dim ddlpf As New DropdownListPropertiesForm()
                    ddlpf.DropDownListWidget = TryCast(clickedFieldWidget, PDFDropDownListWidget)
                    ddlpf.ShowDialog()

                    pageView.EndContentUpdate()
                    Exit Select
                Case PDFFieldType.Signature
                    pageView.BeginContentUpdate()

                    Dim spf As New SignaturePropertiesForm()
                    spf.SignatureWidget = TryCast(clickedFieldWidget, PDFSignatureWidget)
                    spf.ShowDialog()

                    pageView.EndContentUpdate()
                    Exit Select
            End Select
        End Sub

        Private Sub tsmiFieldContextMenuDelete_Click(ByVal sender As Object, ByVal e As EventArgs) Handles tsmiFieldContextMenuDelete.Click
            Dim fieldToDelete As PDFField = clickedFieldWidget.Field
            ' Field has multiple widgets, remove only the selected one.
            If fieldToDelete.Widgets.Count > 1 Then
                fieldToDelete.Widgets.Remove(clickedFieldWidget)
            Else
                ' Remove the complete field.
                document.Form.Fields.Remove(fieldToDelete)
            End If
        End Sub

		#End Region
	End Class
End Namespace

File: WinForms/.NET/FormDesigner/AppForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using O2S.Components.PDFView4NET;
using O2S.Components.PDFView4NET.Forms;

namespace FormDesigner
{
    public partial class AppForm : Form
    {
        public AppForm()
        {
            InitializeComponent();
        }

        private PDFFieldWidget clickedFieldWidget;

        private void AppForm_Load(object sender, EventArgs e)
        {
            tscbxZoom.SelectedIndex = 2;
        }

        #region Toolbar Events

        private void tsbFileOpen_Click(object sender, EventArgs e)
        {
            if (ofd.ShowDialog() == DialogResult.OK)
            {
                document.Load(ofd.FileName);
            }
        }

        private void tsbFileSave_Click(object sender, EventArgs e)
        {
            if (sfd.ShowDialog() == DialogResult.OK)
            {
                document.Save(sfd.FileName);
            }
        }

        private void tsbPanAndScan_Click(object sender, EventArgs e)
        {
            pageView.WorkMode = UserInteractiveWorkMode.PanAndScan;
        }

        private void tsbEditFormFields_Click(object sender, EventArgs e)
        {
            pageView.WorkMode = UserInteractiveWorkMode.EditAnnotations;
        }

        private void tsbFieldTextBox_Click(object sender, EventArgs e)
        {
            pageView.WorkMode = UserInteractiveWorkMode.AddTextBoxField;
        }

        private void tsbFieldCheckBox_Click(object sender, EventArgs e)
        {
            pageView.WorkMode = UserInteractiveWorkMode.AddCheckBoxField;
        }

        private void tsbFieldRadioButton_Click(object sender, EventArgs e)
        {
            pageView.WorkMode = UserInteractiveWorkMode.AddRadioButtonField;
        }

        private void tsbFieldListBox_Click(object sender, EventArgs e)
        {
            pageView.WorkMode = UserInteractiveWorkMode.AddListBoxField;
        }

        private void tsbFieldComboBox_Click(object sender, EventArgs e)
        {
            pageView.WorkMode = UserInteractiveWorkMode.AddDropDownListField;
        }

        private void tsbFieldPushButton_Click(object sender, EventArgs e)
        {
            pageView.WorkMode = UserInteractiveWorkMode.AddPushButtonField;
        }

        private void tsbFieldDigitalSignature_Click(object sender, EventArgs e)
        {
            pageView.WorkMode = UserInteractiveWorkMode.AddSignatureField;
        }

        private void tscbxZoom_SelectedIndexChanged(object sender, EventArgs e)
        {
            pageView.Zoom = float.Parse(tscbxZoom.SelectedItem.ToString());
        }

        #endregion Toolbar Events

        #region PDF Viewer Events

        private void pageView_WorkModeChanged(object sender, EventArgs e)
        {
            tsbPanAndScan.Checked = pageView.WorkMode == UserInteractiveWorkMode.PanAndScan;
            tsbEditFormFields.Checked = pageView.WorkMode == UserInteractiveWorkMode.EditAnnotations;
            tsbFieldTextBox.Checked = pageView.WorkMode == UserInteractiveWorkMode.AddTextBoxField;
            tsbFieldCheckBox.Checked = pageView.WorkMode == UserInteractiveWorkMode.AddCheckBoxField;
            tsbFieldRadioButton.Checked = pageView.WorkMode == UserInteractiveWorkMode.AddRadioButtonField;
            tsbFieldListBox.Checked = pageView.WorkMode == UserInteractiveWorkMode.AddListBoxField;
            tsbFieldComboBox.Checked = pageView.WorkMode == UserInteractiveWorkMode.AddDropDownListField;
            tsbFieldPushButton.Checked = pageView.WorkMode == UserInteractiveWorkMode.AddPushButtonField;
            tsbFieldDigitalSignature.Checked = pageView.WorkMode == UserInteractiveWorkMode.AddSignatureField;
        }

        private void pageView_BeforeFieldAdd(object sender, BeforeFieldAddEventArgs e)
        {
            bool cancel = false;
            switch (e.Field.Type)
            {
                case PDFFieldType.TextBox:
                    TextBoxPropertiesForm tbpf = new TextBoxPropertiesForm();
                    tbpf.TextBoxWidget = e.Field.Widgets[0] as PDFTextBoxWidget;
                    cancel = tbpf.ShowDialog() == DialogResult.Cancel;
                    break;
                case PDFFieldType.CheckBox:
                    CheckBoxPropertiesForm cbpf = new CheckBoxPropertiesForm();
                    cbpf.CheckBoxWidget = e.Field.Widgets[0] as PDFCheckBoxWidget;
                    cancel = cbpf.ShowDialog() == DialogResult.Cancel;
                    break;
                case PDFFieldType.RadioButton:
                    RadioButtonPropertiesForm rbpf = new RadioButtonPropertiesForm();
                    rbpf.RadioButtonWidget = e.Field.Widgets[0] as PDFRadioButtonItem;
                    cancel = rbpf.ShowDialog() == DialogResult.Cancel;
                    break;
                case PDFFieldType.PushButton:
                    PushButtonPropertiesForm pbpf = new PushButtonPropertiesForm();
                    pbpf.PushButtonWidget = e.Field.Widgets[0] as PDFPushButtonWidget;
                    cancel = pbpf.ShowDialog() == DialogResult.Cancel;
                    break;
                case PDFFieldType.ListBox:
                    ListBoxPropertiesForm lbpf = new ListBoxPropertiesForm();
                    lbpf.ListBoxWidget = e.Field.Widgets[0] as PDFListBoxWidget;
                    cancel = lbpf.ShowDialog() == DialogResult.Cancel;
                    break;
                case PDFFieldType.DropDownList:
                    DropdownListPropertiesForm ddlpf = new DropdownListPropertiesForm();
                    ddlpf.DropDownListWidget = e.Field.Widgets[0] as PDFDropDownListWidget;
                    cancel = ddlpf.ShowDialog() == DialogResult.Cancel;
                    break;
                case PDFFieldType.Signature:
                    SignaturePropertiesForm spf = new SignaturePropertiesForm();
                    spf.SignatureWidget = e.Field.Widgets[0] as PDFSignatureWidget;
                    cancel = spf.ShowDialog() == DialogResult.Cancel;
                    break;
            }

            e.Cancel = cancel;
        }

        private void pageView_AfterFieldAdd(object sender, AfterFieldAddEventArgs e)
        {

        }

        private void pageView_FieldContextMenu(object sender, FieldContextMenuEventArgs e)
        {
            if (pageView.WorkMode == UserInteractiveWorkMode.EditAnnotations)
            {
                clickedFieldWidget = e.FieldWidget;
                cmsFieldContextMenu.Show(pageView.PointToScreen(e.MousePosition));
            }
        }

        #endregion PDF Viewer Events

        #region Context Menu Events

        private void tsmiFieldContextMenuProperties_Click(object sender, EventArgs e)
        {
            switch (clickedFieldWidget.Field.Type)
            {
                case PDFFieldType.TextBox:
                    pageView.BeginContentUpdate();

                    TextBoxPropertiesForm tbpf = new TextBoxPropertiesForm();
                    tbpf.TextBoxWidget = clickedFieldWidget as PDFTextBoxWidget;
                    tbpf.ShowDialog();

                    pageView.EndContentUpdate();
                    break;
                case PDFFieldType.CheckBox:
                    pageView.BeginContentUpdate();

                    CheckBoxPropertiesForm cbpf = new CheckBoxPropertiesForm();
                    cbpf.CheckBoxWidget = clickedFieldWidget as PDFCheckBoxWidget;
                    cbpf.ShowDialog();

                    pageView.EndContentUpdate();
                    break;
                case PDFFieldType.RadioButton:
                    pageView.BeginContentUpdate();

                    RadioButtonPropertiesForm rbpf = new RadioButtonPropertiesForm();
                    rbpf.RadioButtonWidget = clickedFieldWidget as PDFRadioButtonItem;
                    rbpf.ShowDialog();

                    pageView.EndContentUpdate();
                    break;
                case PDFFieldType.PushButton:
                    pageView.BeginContentUpdate();

                    PushButtonPropertiesForm pbpf = new PushButtonPropertiesForm();
                    pbpf.PushButtonWidget = clickedFieldWidget as PDFPushButtonWidget;
                    pbpf.ShowDialog();

                    pageView.EndContentUpdate();
                    break;
                case PDFFieldType.ListBox:
                    pageView.BeginContentUpdate();

                    ListBoxPropertiesForm lbpf = new ListBoxPropertiesForm();
                    lbpf.ListBoxWidget = clickedFieldWidget as PDFListBoxWidget;
                    lbpf.ShowDialog();

                    pageView.EndContentUpdate();
                    break;
                case PDFFieldType.DropDownList:
                    pageView.BeginContentUpdate();

                    DropdownListPropertiesForm ddlpf = new DropdownListPropertiesForm();
                    ddlpf.DropDownListWidget = clickedFieldWidget as PDFDropDownListWidget;
                    ddlpf.ShowDialog();

                    pageView.EndContentUpdate();
                    break;
                case PDFFieldType.Signature:
                    pageView.BeginContentUpdate();

                    SignaturePropertiesForm spf = new SignaturePropertiesForm();
                    spf.SignatureWidget = clickedFieldWidget as PDFSignatureWidget;
                    spf.ShowDialog();

                    pageView.EndContentUpdate();
                    break;
            }
        }

        private void tsmiFieldContextMenuDelete_Click(object sender, EventArgs e)
        {
            PDFField fieldToDelete = clickedFieldWidget.Field;
            // Field has multiple widgets, remove only the selected one.
            if (fieldToDelete.Widgets.Count > 1)
            {
                fieldToDelete.Widgets.Remove(clickedFieldWidget);
            }
            else // Remove the complete field.
            {
                document.Form.Fields.Remove(fieldToDelete);
            }
        }

        #endregion Context Menu Events
    }
}
